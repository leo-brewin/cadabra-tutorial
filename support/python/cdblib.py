def create (file_name):
  import json, io, os, errno

  try:
      os.remove(file_name)                # delete the file if it exsists
      with open(file_name, 'w'): pass     # create an empty file
  except OSError as e:
      if e.errno == errno.ENOENT:         # errno.ENOENT = no such file or directory
         with open(file_name, 'w'): pass  # create an empty file
      else:
          raise                           # report an exception

  # Create and save an empty dict
  data_out = {}
  with io.open(os.getcwd() + '/' + file_name, 'w', encoding='utf-8') as out_file:
      out_file.write(json.dumps(data_out,
                                indent=2,
                                sort_keys=True,
                                separators=(',', ': '),
                                ensure_ascii=False)+'\n')

# put and get based on ~/Python/examples/dict-io/lcb-01.py

def put (key_name,object,file_name):
  import json, io, os
  import cadabra2
  __cdbkernel__ = cadabra2.__cdbkernel__

  # Read the current dict
  with io.open(os.getcwd() + '/' + file_name) as inp_file:
      data_out = json.load(inp_file)

  # Add a new entry to the dict.

  # Two choices here. Both give almost identical output except for the rule symbol '->'.
  # Using str(object) will write this using unicode characters. This causes problems
  # down the track -- when these rules are imported in other codes we get a Python
  # error -- it reports that it can't handle this unicode character.
  # Using object.input_form() does the trick. It writes out '->'. This is good.

  # data_out[key_name] = str(object)       # will write rule char using unicode
  data_out[key_name] = object.input_form() # will write rule char as '->'

  # Save the updated dict
  with io.open(os.getcwd() + '/' + file_name, 'w', encoding='utf-8') as out_file:
      out_file.write(json.dumps(data_out,
                                indent=2,
                                sort_keys=True,
                                separators=(',', ': '),
                                ensure_ascii=False)+'\n')

def get (key_name,file_name):
  import json, io, os
  import cadabra2
  __cdbkernel__ = cadabra2.__cdbkernel__

  # Read the current dict
  with io.open(os.getcwd() + '/' + file_name) as inp_file:
      data_inp = json.load(inp_file)

  # Return one entry from the dict
  return cadabra2.Ex (data_inp[key_name])
