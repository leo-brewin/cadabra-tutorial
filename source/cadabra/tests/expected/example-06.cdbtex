\chardef\thehashcode=\catcode`\#	 % save catcode of hash (probably 6)
\catcode`\#=11			             % set catcode for hash to be a letter (11)
\cdbtag{ex-06.100}{%
\left[V_{\theta} = \varphi,~\discretionary{}{}{} V_{\varphi} = \sin{\theta}\right]%
}
\cdbtag{ex-06.101}{%
\partial_{b}{V_{a}}-\partial_{a}{V_{b}}%
}
\cdbtag{ex-06.102}{%
\square{}_{a}{}_{b}\left\{\begin{aligned}\square{}_{\varphi}{}_{\theta}&= \cos{\theta}-1\\[-.5ex]%
\square{}_{\theta}{}_{\varphi}&= -\cos{\theta}+1\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{ex-06.201}{%
\left[g_{\theta \theta} = {r}^{2},~\discretionary{}{}{} g_{\varphi \varphi} = {r}^{2} {\left(\sin{\theta}\right)}^{2}\right]%
}
\cdbtag{ex-06.202}{%
\left[g^{\theta \theta} = {r}^{-2},~\discretionary{}{}{} g^{\varphi \varphi} = {\left({r}^{2} {\left(\sin{\theta}\right)}^{2}\right)}^{-1}\right]%
}
\cdbtag{ex-06.203}{%
\Gamma^{a}{}_{f g} \rightarrow \square{}_{f}{}_{g}{}^{a}\left\{\begin{aligned}\square{}_{\varphi}{}_{\theta}{}^{\varphi}&= {\left(\tan{\theta}\right)}^{-1}\\[-.5ex]%
\square{}_{\theta}{}_{\varphi}{}^{\varphi}&= {\left(\tan{\theta}\right)}^{-1}\\[-.5ex]%
\square{}_{\varphi}{}_{\varphi}{}^{\theta}&=  - \frac{1}{2}\sin\left(2\theta\right)\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{ex-06.204}{%
R^{d}{}_{e f g} \rightarrow \square{}_{e}{}_{g}{}^{d}{}_{f}\left\{\begin{aligned}\square{}_{\varphi}{}_{\varphi}{}^{\theta}{}_{\theta}&= {\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\varphi}{}^{\varphi}{}_{\theta}&= -1\\[-.5ex]%
\square{}_{\varphi}{}_{\theta}{}^{\theta}{}_{\varphi}&= -{\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\theta}{}^{\varphi}{}_{\varphi}&= 1\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{ex-06.205}{%
R_{a b} \rightarrow \square{}_{a}{}_{b}\left\{\begin{aligned}\square{}_{\varphi}{}_{\varphi}&= {\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\theta}&= 1\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{ex-06.206}{%
\square{}_{e}{}_{g}{}^{d}{}_{f}\left\{\begin{aligned}\square{}_{\varphi}{}_{\varphi}{}^{\theta}{}_{\theta}&= {\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\varphi}{}^{\varphi}{}_{\theta}&= -1\\[-.5ex]%
\square{}_{\varphi}{}_{\theta}{}^{\theta}{}_{\varphi}&= -{\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\theta}{}^{\varphi}{}_{\varphi}&= 1\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{print.ex-06.02}{%
Rab = R_{a b} -> \components_{a b}({{\varphi, \varphi} = (\sin(\theta))**2, {\theta, \theta} = 1})%
Rab[0] = R_{a b}%
Rab[1] = \components_{a b}({{\varphi, \varphi} = (\sin(\theta))**2, {\theta, \theta} = 1})%
Rab[1][0] = a%
Rab[1][1] = b%
Rab[1][2] = {{\varphi, \varphi} = (\sin(\theta))**2, {\theta, \theta} = 1}%
Rab[1][2][0] = {\varphi, \varphi} = (\sin(\theta))**2%
Rab[1][2][0][0] = {\varphi, \varphi}%
Rab[1][2][0][1] = (\sin(\theta))**2%
}
\cdbtag{ex-06.301}{%
\left[g_{\theta \theta} = {r}^{2},~\discretionary{}{}{} g_{\varphi \varphi} = {r}^{2} {\left(\sin{\theta}\right)}^{2}\right]%
}
\cdbtag{ex-06.302}{%
\left[g_{\theta \theta} = {r}^{2},~\discretionary{}{}{} g_{\varphi \varphi} = {r}^{2} {\left(\sin{\theta}\right)}^{2},~\discretionary{}{}{} g^{\theta \theta} = {r}^{-2},~\discretionary{}{}{} g^{\varphi \varphi} = {\left({r}^{2} {\left(\sin{\theta}\right)}^{2}\right)}^{-1}\right]%
}
\cdbtag{ex-06.303}{%
\Gamma^{a}{}_{f g} \rightarrow \square{}_{f}{}_{g}{}^{a}\left\{\begin{aligned}\square{}_{\varphi}{}_{\theta}{}^{\varphi}&= {\left(\tan{\theta}\right)}^{-1}\\[-.5ex]%
\square{}_{\theta}{}_{\varphi}{}^{\varphi}&= {\left(\tan{\theta}\right)}^{-1}\\[-.5ex]%
\square{}_{\varphi}{}_{\varphi}{}^{\theta}&=  - \frac{1}{2}\sin\left(2\theta\right)\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{ex-06.304}{%
R^{d}{}_{e f g} \rightarrow \square{}_{e}{}_{g}{}^{d}{}_{f}\left\{\begin{aligned}\square{}_{\varphi}{}_{\varphi}{}^{\theta}{}_{\theta}&= {\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\varphi}{}^{\varphi}{}_{\theta}&= -1\\[-.5ex]%
\square{}_{\varphi}{}_{\theta}{}^{\theta}{}_{\varphi}&= -{\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\theta}{}^{\varphi}{}_{\varphi}&= 1\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{ex-06.305}{%
R_{a b} \rightarrow \square{}_{a}{}_{b}\left\{\begin{aligned}\square{}_{\varphi}{}_{\varphi}&= {\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\theta}&= 1\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{ex-06.306}{%
\square{}_{e}{}_{g}{}^{d}{}_{f}\left\{\begin{aligned}\square{}_{\varphi}{}_{\varphi}{}^{\theta}{}_{\theta}&= {\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\varphi}{}^{\varphi}{}_{\theta}&= -1\\[-.5ex]%
\square{}_{\varphi}{}_{\theta}{}^{\theta}{}_{\varphi}&= -{\left(\sin{\theta}\right)}^{2}\\[-.5ex]%
\square{}_{\theta}{}_{\theta}{}^{\varphi}{}_{\varphi}&= 1\\[-.5ex]%
\end{aligned}\right.%
}
\cdbtag{ex-06.400}{%
\left[g_{\theta \theta} = {r}^{2},~\discretionary{}{}{} g_{\varphi \varphi} = {r}^{2} {\left(\sin{\theta}\right)}^{2}\right]%
}
\cdbtag{ex-06.401}{%
g_{a b} \varphi^{a} \varphi^{b}%
}
\cdbtag{ex-06.402}{%
{r}^{2} {\left(\sin{\theta}\right)}^{2}%
}
\cdbtag{print.ex-06.04}{%
type compt   = <class 'cadabra2.Ex'>%
type gphiphi = <class 'sympy.core.mul.Mul'>%
     compt   = (r)**2 (\sin(\theta))**2%
     gphiphi = r**2*sin(theta)**2%
}
\cdbtag{ex-06.500}{%
\left[g_{\theta \theta} = {r}^{2},~\discretionary{}{}{} g_{\varphi \varphi} = {r}^{2} {\left(\sin{\theta}\right)}^{2}\right]%
}
\cdbtag{ex-06.501}{%
\left[\varphi,~\discretionary{}{}{} \varphi\right]%
}
\cdbtag{ex-06.502}{%
{r}^{2} {\left(\sin{\theta}\right)}^{2}%
}
\cdbtag{print.ex-06.05}{%
metric = \components_{a b}({{\theta, \theta} = (r)**2, {\varphi, \varphi} = (r)**2 (\sin(\theta))**2})%
metric[0] = a%
metric[1] = b%
metric[2] = {{\theta, \theta} = (r)**2, {\varphi, \varphi} = (r)**2 (\sin(\theta))**2}%
metric[2][1] = {\varphi, \varphi} = (r)**2 (\sin(\theta))**2%
metric[2][1][0] = {\varphi, \varphi}%
metric[2][1][1] = (r)**2 (\sin(\theta))**2%
}
\catcode`\#=\thehashcode          % restore original catcode for hash
