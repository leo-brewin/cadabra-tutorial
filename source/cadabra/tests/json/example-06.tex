\documentclass[12pt]{cdblatex}

\begin{document}

% ============================================================================================
\section*{Example 6 Evaluating components}

\begin{cadabra}
   {\theta, \varphi}::Coordinate.
   {a,b,c,d,e,f,g,h#}::Indices(values={\theta, \varphi}, position=independent).

   \partial{#}::PartialDerivative.

   def check (objA,objB):

       tmp := @(objA) - @(objB).

       distribute     (tmp)
       sort_product   (tmp)
       rename_dummies (tmp)
       canonicalise   (tmp)

       return tmp

   import cdblib

   # ---------------------------------------------
   json000A = cdblib.get ('json000','../expected/example-06.json')
   json001A = cdblib.get ('json001','../expected/example-06.json')
   json002A = cdblib.get ('json002','../expected/example-06.json')

   # ---------------------------------------------
   json000B = cdblib.get ('json000','../output/example-06.json')
   json001B = cdblib.get ('json001','../output/example-06.json')
   json002B = cdblib.get ('json002','../output/example-06.json')

   # ---------------------------------------------
   diff000 = check (json000A,json000B)   # cdb (ex-06.diff000,diff000)
   diff001 = check (json001A,json001B)   # cdb (ex-06.diff001,diff001)
   diff002 = check (json002A,json002B)   # cdb (ex-06.diff002,diff002)

\end{cadabra}

\clearpage

\begin{dgroup*}
   \Dmath*{\cdb*{ex-06.diff000}}
   \Dmath*{\cdb*{ex-06.diff001}}
   \Dmath*{\cdb*{ex-06.diff002}}
\end{dgroup*}

\end{document}
