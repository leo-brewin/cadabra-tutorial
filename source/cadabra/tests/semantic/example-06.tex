\def\Date{10 Aug 2022}

\documentclass[12pt]{cdblatex}
\usepackage{examples}

\begin{document}

% ============================================================================================
\section*{Example 6 Evaluating components}

\CdbSetup{action=hide}

\begin{cadabra}
   {\theta, \varphi}::Coordinate.
   {a,b,c,d,e,f,g,h#}::Indices(values={\theta, \varphi}, position=independent).

   \partial{#}::PartialDerivative.

   def check (objA,objB):

       tmp := @(objA) - @(objB).

       distribute     (tmp)
       sort_product   (tmp)
       rename_dummies (tmp)
       canonicalise   (tmp)

       return tmp

   import cdblib

   # ---------------------------------------------
   check000A = cdblib.get ('check000','expected/example-06.json')
   check001A = cdblib.get ('check001','expected/example-06.json')
   check002A = cdblib.get ('check002','expected/example-06.json')

   # ---------------------------------------------
   check000B = cdblib.get ('check000','output/example-06.json')
   check001B = cdblib.get ('check001','output/example-06.json')
   check002B = cdblib.get ('check002','output/example-06.json')

   # ---------------------------------------------
   diff000 = check (check000A,check000B)   # cdb (ex-06.diff000,diff000)
   diff001 = check (check001A,check001B)   # cdb (ex-06.diff001,diff001)
   diff002 = check (check002A,check002B)   # cdb (ex-06.diff002,diff002)

\end{cadabra}

% \clearpage

\begin{dgroup*}
   \Dmath*{\cdb*{ex-06.diff000}}
   \Dmath*{\cdb*{ex-06.diff001}}
   \Dmath*{\cdb*{ex-06.diff002}}
\end{dgroup*}

\end{document}
