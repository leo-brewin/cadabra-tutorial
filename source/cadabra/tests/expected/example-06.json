{
  "json000": "(r)**2 (\\sin(\\theta))**2",
  "json001": "{\\varphi, \\varphi}",
  "json002": "(r)**2 (\\sin(\\theta))**2"
}
